internal/controller/controllers/spoke_k8s_client.go

```diff
@@ -0,0 +1,84 @@
+package controllers
+
+import (
+	"context"
+
+	"github.com/sirupsen/logrus"
+	certificatesv1 "k8s.io/api/certificates/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	cerv1 "k8s.io/client-go/kubernetes/typed/certificates/v1"
+	typedcorev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+)
+
+//go:generate mockgen -package=controllers -destination=mock_spoke_k8s_client_factory.go . SpokeK8sClientFactory
+type SpokeK8sClientFactory interface {
+	Create(secret *corev1.Secret) (SpokeK8sClient, error)
+}
+
+//go:generate mockgen -package=controllers -destination=mock_spoke_k8s_client.go . SpokeK8sClient
+type SpokeK8sClient interface {
+	ListCsrs() (*certificatesv1.CertificateSigningRequestList, error)
+	ApproveCsr(csr *certificatesv1.CertificateSigningRequest) error
+	GetNode(name string) (*corev1.Node, error)
+}
+
+type spokeK8sClient struct {
+	csrClient   cerv1.CertificateSigningRequestInterface
+	nodesClient typedcorev1.NodeInterface
+	log         logrus.FieldLogger
+}
+
+type spokeK8sClientFactory struct {
+	log logrus.FieldLogger
+}
+
+func NewSpokeK8sClientFactory(log logrus.FieldLogger) SpokeK8sClientFactory {
+	return &spokeK8sClientFactory{
+		log: log,
+	}
+}
+
+func (cf *spokeK8sClientFactory) Create(secret *corev1.Secret) (SpokeK8sClient, error) {
+	clientConfig, err := getRestConfig(secret)
+	if err != nil {
+		cf.log.WithError(err).Warnf("Getting client from kubeconfig cluster")
+		return nil, err
+	}
+	config, err := kubernetes.NewForConfig(clientConfig)
+	if err != nil {
+		cf.log.WithError(err).Warnf("Getting kuberenetes config for cluster")
+		return nil, err
+	}
+	data := spokeK8sClient{
+		csrClient:   config.CertificatesV1().CertificateSigningRequests(),
+		nodesClient: config.CoreV1().Nodes(),
+		log:         cf.log,
+	}
+	return &data, nil
+}
+
+func (c *spokeK8sClient) ListCsrs() (*certificatesv1.CertificateSigningRequestList, error) {
+	return c.csrClient.List(context.TODO(), metav1.ListOptions{})
```

Comment:
Let's filter only for `Pending` CSRs. Normally, when CSRs are `Approved` they are garbage collected but it takes a bit.

Reply:
To find if CSR is pending, we need to filter the CSR to find if "Approved" condition does not exist.  @flaper87 do you know if it is possible and how to do it?

Reply:
Putting it on hold for now